packages/db/src/query/peaks.ts

```diff
@@ -0,0 +1,308 @@
+    while (i < values.length) {
+        const entry = values[i];
+
+        if (entry.value > threshold) {
+            const isStart = i === 0;
+            let sequenceEnd = i + 1;
+
+            while (sequenceEnd < values.length && values[sequenceEnd].value > threshold) {
+                sequenceEnd++;
+            }
+
+            const sequenceLength = sequenceEnd - i;
+
+            // only mark as peak if longer then 2min and not marked as anomaly yet
+            if (sequenceLength > 5) {
+                const avgPeakPower = calculateAveragePower(values.slice(i, sequenceEnd));
+
+                sequences.push({
+                    start: entry.timestamp,
+                    end: values[sequenceEnd - 1].timestamp,
+                    isAtStart: isStart,
+                    averagePowerIncludingBaseLoad: avgPeakPower,
+                });
+            }
+            i = sequenceEnd;
+        } else {
+            i++;
+        }
+    }
+
+    return sequences;
+}
+
+/**
+ * Finds peaks in a given set of values
+ * for peaks 5 is the default threshold, for anomalies 1000 is a good fit
+ */
+export function findPeaks(
+    thresholdValues: SensorDataSelectType[],
+    consideredValues: SensorDataSelectType[],
+    threshold = 5,
+    medianValue?: number,
+) {
+    let median: number | undefined = medianValue;
+    if (!median) {
+        median = calculateMedian(thresholdValues);
+    }
+    const madValue = calculateMAD(thresholdValues, 1.4826, median);
+    const processedValues = consideredValues.map((value) => ({
+        ...value,
+        value: Math.abs(value.value - median),
+    }));
+    return findSequences(processedValues, madValue * threshold);
+}
+
+interface FindAndMarkPeaksProps {
+    sensorId: string;
+    start: Date;
+    end: Date;
+    type: "peak" | "anomaly";
+}
+
+export async function findAndMark(props: FindAndMarkPeaksProps, threshold = 5) {
+    const { sensorId, start, end, type } = props;
+
+    // we shift the start 12 hours back, so we have a bigger sample for the threshold
+    const sequenceStart = new Date(start);
+    sequenceStart.setHours(sequenceStart.getHours() - 24, 0, 0, 0);
+
+    try {
+        return await db.transaction(async (trx) => {
+            const calcDbData = await trx
+                .select()
+                .from(sensorData)
+                .where(and(eq(sensorData.sensorId, sensorId), between(sensorData.timestamp, sequenceStart, end)))
+                .orderBy(asc(sensorData.timestamp));
+
+            // make sure we have at least 3 hours of reference data
+            if (calcDbData.length === 0 || calcDbData.length < 720) {
+                return [];
+            }
+
+            const calcData = calcDbData
+                .map((d, i) => {
+                    return {
+                        ...d,
+                        value: i === 0 ? 0 : Number(d.value) - Number(calcDbData[i - 1].value),
+                    };
+                })
+                .slice(1);
```

Comment:
Können wir hier den query aus dem draft der Strom Seite nehmen? #138

Ist in sensor.ts dort.
Wir sparen uns das Mapping und den slice.

Reply:
Habe ich erst mal so rein kopiert. Sollten wir dann später meiner Meinung nach in eine einheitliche Methode auslagern, die wir an den verschiedenen Stellen aufrufen können :+1: